README.md
## Flat-Table: Dictionary and List Normalizer

This package is a normalizer for [pandas](https://pandas.pydata.org/) dataframe objects that has dictionary or list objects within it's columns. The library will expand all of the columns that has data types in (list, dict) into individual seperate rows and columns.

Flat table will use *the current index* of the dataframe as an identifier while expanding lists. The output will have an index column of your original dataframe. You can drop it later if you not plan to use it.

For changes between versions, check out the [changelog](CHANGELOG.md).

### To Install

To install, use pip.

```
python -m pip install flat-table
```

### How to Use It

From a given pandas dataframe, the `index` of the dataframe will be used to create seperate columns and rows. 

``` python
# some dataframe contains dicts and lists in it's columns
df = ...
```

``` python
import flat_table

flat_table.normalize(df)
```

This will give you all the keys in dictionaries as columns, and all the lists as seperate rows.

### Example Illustration

Lets assume that you have a dataframe of the followings shape.


id | user_info | address
-- | --------- | ------ |
1001 | { 'first_name': 'john', 'last_name': 'smith', 'phones': {'mobile': '201-..', 'home': '978-..'} }| [{ 'zip': '07014', 'city': 'clifton' }] |
1002 | NaN| [{'zip': '07014', 'address1': '1 Journal Square'}]|
1003 | { 'first_name': 'marry', 'last_name': 'kate', 'gender': 'female'  } | [{ 'zip': '10001', 'city': 'new york' }, { 'zip': '10008', 'city': 'brooklyn' }]|


This table given above has some dictionaries and lists in it's columns. Normally, what you would do is to use `pd.io.json.json_normalize` function to expand dictionaries. However, in cases you have `NaN` values in your column, `pd.io.json.json_normalize` end up throwing an `AttributeError` error for `NaN` values because they are not of the same type. `flat_table` is a wraper around the `json_normalize` function where it expands it's abilities to be more robust for `NaN` values and also, it expands lists rowwise so that it will be more clear to see the information.

For the above table, the flatten table after applying `flat_table.normalize` will look like the following.

|    |   index |   id | user_info.gender   | user_info.phones.home   | user_info.phones.mobile   | user_info.last_name   | user_info.first_name   | address.address1   | address.city   |   address.zip |
|---:|--------:|-----:|:-------------------|:------------------------|:--------------------------|:----------------------|:-----------------------|:-------------------|:---------------|--------------:|
|  0 |       0 | 1001 | nan                | 978-..                  | 201-..                    | smith                 | john                   | nan                | clifton        |         07014 |
|  1 |       1 | 1002 | nan                | nan                     | nan                       | nan                   | nan                    | 1 Journal Square   | nan            |         07014 |
|  2 |       2 | 1003 | female             | nan                     | nan                       | kate                  | marry                  | nan                | new york       |         10001 |
|  3 |       2 | 1003 | female             | nan                     | nan                       | kate                  | marry                  | nan                | brooklyn       |         10008 |

The expansion for dicts and lists made optional. If not needed, you can choose to expand list types and dict types with normalize function.

``` python
flat_table.normalize(df, expand_dicts=False, expand_lists=True)
```

Normalized version of df will be following.

|    |   index |   id | user_info                                                                                      | address.address1   | address.city   |   address.zip |
|---:|--------:|-----:|:-----------------------------------------------------------------------------------------------|:-------------------|:---------------|--------------:|
|  0 |       0 | 1001 | {...} | nan                | clifton        |         07014 |
|  1 |       1 | 1002 | nan                                                                                            | 1 Journal Square   | nan            |         07014 |
|  2 |       2 | 1003 | {...}                               | nan                | new york       |         10001 |
|  3 |       2 | 1003 | {...}                               | nan                | brooklyn       |         10008 |

### How it Works?

`flat_table` will look for each of the series in a dataframe to understand what type of data it contains. 

For every series, it creates a list of information on how to expand it. It will go into all dictionaries and all lists in all levels and expand them as rows and columns. Dictionary `keys` will be used for column names, and The `index` of the giden dataframe will be used for row expansion.

If you want to see how the columns are mapped, you can use `flat_table.mapper` function to get all information about your columns in your original dataframe. For example, for the above table, the mapper function will provide the following table.

|    | parent    | child                   | type   | obj   |
|---:|:----------|:------------------------|:-------|:------|
|  0 | .         | id                      | int    | ...   |
|  1 | .         | user_info               | dict   | ...   |
|  2 | user_info | user_info.gender        | str    | ...   |
|  3 | user_info | user_info.phones.home   | str    | ...   |
|  4 | user_info | user_info.phones.mobile | str    | ...   |
|  5 | user_info | user_info.last_name     | str    | ...   |
|  6 | user_info | user_info.first_name    | str    | ...   |
|  7 | .         | address                 | list   | ...   |
|  8 |           | address                 | dict   | ...   |
|  9 | address   | address.address1        | str    | ...   |
| 10 | address   | address.city            | str    | ...   |
| 11 | address   | address.zip             | str    | ...   |

## Licence

Licence is use it at your own will, with whatever way you want it to use :smiley:.

## Author

Build by [@metinsenturk](https://github.com/metinsenturk/).


CHANGELOG.md
# Changelog for Flat Table!

Changes of things for this package between versions is listed in below.

# 1.1.2 - 2020-08-31

- Added typing hints.

# 1.1.1 - 2020-08-23

- Updated pandas to latest (1.1.1).
- Using numpy from pandas is deprecated. Added numpy as seperate package.
- Updated notebooks to make it compatible with pandas 1.1.x versions.
- Updated the requirements file with only the packages needed, removed dependencies.

# 1.1.0 - 2019-10-16

- The expansion for dicts and lists made optional.

# 1.0.0 - 2019-09-26

- Initial release.


norm.py
#!/usr/bin/python3
from functools import reduce
from itertools import chain
import pandas as pd
import datetime
import logging
import os

now = datetime.datetime.now()
dirname = os.path.dirname(os.path.realpath('__file__'))
logfile = "{0}/{1}/{2}.log".format(dirname, 'logs', now.strftime(r"%y-%m-%d-%H%M"))

logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    level=logging.INFO,
    handlers=[
        logging.FileHandler(filename=logfile, mode='a+'),
        logging.StreamHandler()
    ])

logger = logging.getLogger(__name__)


def get_list_obj(series):
    """ returns a dataseries of given series object. expands by the rows  """
    ds = series.copy()
    df = ds.reset_index()
    data = df.apply(
        lambda x: [
            {'index': x['index'], ds.name: ins} 
            for ins in x[ds.name]
        ] 
        if type(x[ds.name]) == list 
        else [{'index': x['index'], ds.name: x[ds.name]}],
        axis=1
    )
    df = pd.DataFrame(list(chain(*data)))

    return set_index(df).iloc[:, 0]


def to_index(series):
    """ returns dataframe with index and the series.name """
    df_obj = series.reset_index()
    return df_obj


def set_index(df_or_ds, index_values=None):
    """ utility function to re set indexes in series and dataframes """
    temp = df_or_ds.copy()
    if index_values is None:
        temp.index = temp['index'].values
        temp.drop('index', inplace=True, axis=1)
    else:
        temp.index = index_values
    return temp


def to_columns(series):
    """ returns dataframe of normalized dictionary object """
    ds = series.copy()
    # rows with values
    ds_withvalues = ds[pd.notna(ds)]
    df_withvalues = pd.io.json.json_normalize(ds_withvalues)
    df_withvalues.index = ds_withvalues.index

    # rows with nans
    ds_withnans = ds[pd.isna(ds)]
    df_withnans = pd.DataFrame(columns=df_withvalues.columns, index=ds_withnans.index)

    # concat both
    df = pd.concat([df_withvalues, df_withnans]).sort_index()
    return df


def to_rows(series):
    """ gets the object from a list in a series """
    ds = series
    while True:
        ds = get_list_obj(ds)
        inside = ds.apply(lambda x: type(x)).iloc[0]
        logger.info(f'<Trial #  type: {inside} , len: {len(ds)}>')
        if inside != list:
            break

    return ds


def print_parent_child_node(parent: str, child: pd.Series):
    """ shows parent child relation btw columns and their values """
    logger.info('{:40} {:20} {:15} {:10}'.format(
        parent, child.name, get_type(child), str(child.shape)))


def get_type(child):
    """ get type of the values of the dataseries object """
    df = child[pd.notna(child)]
    typ = ''
    if len(df) > 0:
        typ = df.apply(lambda x: type(x)).iloc[0].__name__
    else:
        typ = type(pd.np.nan).__name__
    
    return typ


def mapper(df):
    """ maps the relationship rowwise and columnwise expansion. """
    headers = ['parent', 'child', 'type', 'obj']
    series_list = [('.', n, get_type(s), s) for n, s in df.iteritems()]

    for ind, (parent, name, typ, child) in enumerate(series_list):
        inside = get_type(child)
        name = ''.join([parent, name]) if parent == '.' else '.'.join([parent, name])

        # parent child nodes '.' level
        print_parent_child_node(parent, child)

        # expand rowwise, add new series for processing
        if inside == 'list':
            _child = to_rows(child)
            series_list.insert(
                ind + 1, (name, _child.name, get_type(_child), _child))
            print_parent_child_node(name, _child)

        # expand columnwise, add new columns for processing
        if inside == 'dict':
            temp = to_columns(child)
            if temp.shape[1] > 1:
                for _name, _child in temp.iteritems():
                    series_list.insert(
                        ind + 1, (name, _name, get_type(_child), _child))
                    print_parent_child_node(name, _child)
            else:
                _child = temp.iloc[:, 0]
                series_list.insert(
                    ind + 1, (name, _child.name, get_type(_child), _child))
                print_parent_child_node(name, _child)

    return pd.DataFrame(data=series_list, columns=headers)


def normalize(df):
    """ normalize rows and columns of a given dataframe """
    dfs = []
    rts = []
    dataframe = df[(df.type != 'dict') & (df.type != 'list')]
    # final packing for list type childs  (concat)
    for parent in dataframe.parent.unique():
        group = dataframe[dataframe.parent.isin([parent])]
        df_group = pd.concat([i for i in group.obj], axis=1)
        dfs.append(df_group)
        logger.info('{:40} before: {:7} after: {:7} obj.shape: {:7} columns: {}'.format(
            parent, 
            str(group.shape),
            str(df_group.shape),
            str(group.obj.iloc[0].shape),
            str(', '.join(df_group.columns.tolist())[:50] + '..')
        ))

    # final packing for all childs        (merge)
    for _df in dfs:
        rts.append(to_index(_df))

    return reduce(lambda x,y: pd.merge(x, y, on='index'), rts)

flat_table/_norm.py
import logging
from typing import List, Any, Union
from functools import reduce
from itertools import chain

import pandas as pd
import numpy as np

__all__ = ['mapper', 'normalize']

logger = logging.getLogger('flat_table')


def get_obj_from_iterable(series: pd.Series) -> pd.Series:
    """ returns a dataseries of given series object. expands by the rows  """
    ds = series.copy()
    df = ds.reset_index()
    data = df.apply(
        lambda x: [
            {'index': x['index'], ds.name: ins}
            for ins in x[ds.name]
        ]
        if type(x[ds.name]) == list
        else [{'index': x['index'], ds.name: x[ds.name]}],
        axis=1
    )
    df = pd.DataFrame(list(chain(*data)))

    return set_index(df).iloc[:, 0]


def to_index(series: pd.Series) -> pd.DataFrame:
    """ returns dataframe with index and the series.name """
    df_obj = series.reset_index()
    return df_obj


def set_index(df_or_ds: Union[pd.DataFrame, pd.Series], index_values: List[Any] = None) -> pd.DataFrame:
    """ utility function to re set indexes in series and dataframes """
    temp = df_or_ds.copy()
    if index_values is None:
        temp.index = temp['index'].values
        temp.drop('index', inplace=True, axis=1)
    else:
        temp.index = index_values
    return temp


def to_columns(series: pd.Series) -> pd.DataFrame:
    """ returns dataframe of normalized dictionary object """
    ds = series.copy()
    # rows with values
    ds_withvalues = ds[pd.notna(ds)]
    df_withvalues = pd.json_normalize(ds_withvalues)
    df_withvalues.index = ds_withvalues.index

    # rows with nans
    ds_withnans = ds[pd.isna(ds)]
    df_withnans = pd.DataFrame(
        columns=df_withvalues.columns, index=ds_withnans.index)

    # concat both
    df = pd.concat([df_withvalues, df_withnans]).sort_index()
    return df


def to_rows(series: pd.Series) -> pd.Series:
    """ gets the object from a list in a series """
    ds = series
    while True:
        ds = get_obj_from_iterable(ds)
        inside = ds.apply(lambda x: type(x)).iloc[0]
        logger.info(f'<Trial #  type: {inside} , len: {len(ds)}>')
        if inside != list:
            break

    return ds


def print_parent_child_node(parent: str, child: pd.Series) -> None:
    """ shows parent child relation btw columns and their values """
    logger.info('{:40} {:20} {:15} {:10}'.format(
        parent, child.name, get_type(child), str(child.shape)))


def get_type(child: pd.Series) -> str:
    """ get type of the values of the dataseries object """
    df = child[pd.notna(child)]
    typ = ''
    if len(df) > 0:
        typ = df.apply(lambda x: type(x)).iloc[0].__name__
    else:
        typ = type(np.nan).__name__

    return typ


def mapper(df: pd.DataFrame) -> pd.DataFrame:
    """
    Maps the relationship rowwise and columnwise expansion.

    Params:
    ------------
    df : a pandas dataframe object of your dataset.
    """
    headers = ['parent', 'child', 'type', 'obj']
    series_list = [('.', n, get_type(s), s) for n, s in df.iteritems()]

    for ind, (parent, name, _, child) in enumerate(series_list):
        inside = get_type(child)

        # parent name
        if parent == '.' or parent == '':
            parent_name = name
        else:
            parent_name = '.'.join([parent, name])

        # parent child nodes '.' level
        print_parent_child_node(parent, child)

        def insert_to_series(p_name: str, _child: pd.Series):
            """ helper to insert child into series_list. """
            if p_name == '.' or p_name == '':
                c_name = _child.name
            else:
                c_name = '.'.join((p_name, _child.name))
            _child = _child.rename(c_name)
            series_list.insert(
                ind + 1, (p_name, c_name, get_type(_child), _child))
            print_parent_child_node(p_name, _child)

        # expand rowwise, add new series for processing
        if inside == 'list':
            _child = to_rows(child)
            insert_to_series('', _child)

        # expand columnwise, add new columns for processing
        if inside == 'dict':
            temp = to_columns(child)
            if temp.shape[1] > 1:
                for _name, _child in temp.iteritems():
                    insert_to_series(parent_name, _child)
            else:
                _child = temp.iloc[:, 0]
                insert_to_series(parent_name, _child)

    return pd.DataFrame(data=series_list, columns=headers)


def normalize(df: pd.DataFrame, expand_dicts: bool = True, expand_lists: bool = True,
              is_mapper: bool = False) -> pd.DataFrame:
    """
    Normalize rows and columns of a given dataframe.

    Params:
    ------------
    df : a pandas dataframe object. It can be your dataframe of
        your dataset, or the output of the mapper() function.
    is_mapper : a boolean to set if you want to use mapper()
        function's dataframe.
    """
    # func init
    if is_mapper:
        headers = ['parent', 'child', 'type', 'obj']
        if all(df.columns != headers):
            raise Exception('Dataframe is not a type of mapper.')
        mp = df
    else:
        mp = mapper(df)

    # option settings
    if expand_dicts and expand_lists:
        dataframe = mp[(mp.type != 'dict') & (mp.type != 'list')]
    # dicts expand, lists are same
    elif expand_dicts and not expand_lists:
        org_df = mp[mp.parent == '.']
        dict_items = org_df[org_df.type == 'dict'].child
        final_dict_items = mp[mp.parent.isin(dict_items)]
        final_others = org_df[org_df.type != 'dict']
        dataframe = pd.concat([final_dict_items, final_others]).sort_index()
    # dicts are same, lists expand
    elif not expand_dicts and expand_lists:
        org_df = mp[mp.parent == '.']
        list_items = mp[mp.type == 'list'].child
        final_list_items = mp[
            (mp.parent.isin(list_items) | mp.child.isin(list_items))
            & ~mp.type.isin(['dict', 'list'])]
        final_others = org_df[org_df.type != 'list']
        dataframe = pd.concat([final_list_items, final_others]).sort_index()
    # original df
    else:
        dataframe = mp[mp.parent == '.']

    # final packing for list type childs  (concat)
    dfs: List[pd.DataFrame] = []
    rts: List[pd.DataFrame] = []
    for parent in dataframe.parent.unique():
        group = dataframe[dataframe.parent.isin([parent])]
        df_group = pd.concat([i for i in group.obj], axis=1)
        dfs.append(df_group)
        logger.info('{:40} before: {:7} after: {:7} obj: {:7} col: {}'.format(
            parent,
            str(group.shape),
            str(df_group.shape),
            str(group.obj.iloc[0].shape),
            str(', '.join(df_group.columns.tolist())[:50] + '..')
        ))

    # final packing for all childs        (merge)
    for _df in dfs:
        rts.append(to_index(_df))

    return reduce(lambda x, y: pd.merge(x, y, on='index'), rts)
